force-app/main/service/DonationHistoryService.cls

```diff
@@ -0,0 +1,176 @@
+
+    static final String HARD_CREDIT = 'Hard Credit';
+    static final String SOFT_CREDIT = 'Soft Credit';
+    static final String PARTIAL_SOFT_CREDIT = 'Partial Soft Credit';
+    
+    /** 
+    * @author Salesforce.org
+    * @date 2021
+    * @description method to return Donation History list.
+    */
+    public List<DonationHistoryController.DonationHistory> getDonationHistory(Id contactId){
+        List<OpportunityContactRole> donations = [SELECT Id, Opportunity.Name, Opportunity.CloseDate, Opportunity.Amount, Contact.Email, Contact.Name, Role, isPrimary 
+                                                    FROM OpportunityContactRole 
+                                                    WHERE Opportunity.IsWon = true 
+                                                    AND ContactId =:contactId 
+                                                    ORDER BY Opportunity.CloseDate 
+                                                    DESC LIMIT 2000];
+        List<DonationHistoryController.DonationHistory> result = verifyCreditCondition(donations);
+        return result;
+    }
+
+    /** 
+    * @author Salesforce.org
+    * @date 2021
+    * @description method to return total number of record
+    */
+    public static List<AggregateResult> getTotalRecords(Id contactId){
+        List<AggregateResult> totalRecords = [SELECT COUNT(Id)
+                                FROM OpportunityContactRole 
+                                WHERE Opportunity.IsWon = true 
+                                AND ContactId =:contactId];
+        return totalRecords;
+    }
+
+    /** 
+    * @author Salesforce.org
+    * @date 2021
+    * @description Verify each condition to list Donation.
+    */
+    private static List<DonationHistoryController.DonationHistory> verifyCreditCondition(List<OpportunityContactRole> donations){
+        List<DonationHistoryController.DonationHistory> donationHistory = new List<DonationHistoryController.DonationHistory>();
+        List<DonationHistoryController.DonationHistory> hardCredit = verifyHardCreditCondition(donations);
+        List<DonationHistoryController.DonationHistory> softCredit = verifySoftCreditCondition(donations);
+        List<Id> donationsId = filterVerifiedIds(softCredit);
+        List<DonationHistoryController.DonationHistory> partialSoftCredit = verifyPartialSoftCreditCondition(donationsId, donations);
+        List<DonationHistoryController.DonationHistory> partialAndSoftCredit = verifyPartialAndSoftCreditCondition(softCredit, partialSoftCredit);
+        
+        donationHistory.addAll(hardCredit);
+        donationHistory.addAll(partialAndSoftCredit);
+
+        return donationHistory;
+    }
+    
+    /** 
+    * @author Salesforce.org
+    * @date 2021
+    * @description Verify hard credit condition.
+    */
+    private static List<DonationHistoryController.DonationHistory> verifyHardCreditCondition(List<OpportunityContactRole> donations){
```

Comment:
The name of this function could describe it better, "verify" isn't what we're doing here. I think a better name here would be `buildHardCreditHistoryFromOCRs`.